src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/transport/OpenDistroPortUnificationHandler.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Portions Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazon.opendistroforelasticsearch.security.ssl.transport;
+
+
+import com.amazon.opendistroforelasticsearch.security.ssl.OpenDistroSecurityKeyStore;
+import com.amazon.opendistroforelasticsearch.security.ssl.util.SSLConnectionTestUtil;
+import com.amazon.opendistroforelasticsearch.security.ssl.util.SSLUtil;
+import com.google.common.annotations.VisibleForTesting;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import java.nio.charset.StandardCharsets;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.net.ssl.SSLException;
+import java.util.List;
+
+/**
+ * Modifies the current pipeline dynamically to enable TLS
+ */
+public class OpenDistroPortUnificationHandler extends ByteToMessageDecoder {
+
+    private final OpenDistroSecurityKeyStore odsks;
+    private static final Logger logger = LogManager.getLogger(OpenDistroPortUnificationHandler.class);
+
+    private SslHandler providedSSLHandler;
+    private SSLUtil sslUtils;
+
+    public OpenDistroPortUnificationHandler(OpenDistroSecurityKeyStore odsks, SSLUtil sslUtils) {
+        this(odsks, null, sslUtils);
+    }
+
+    @VisibleForTesting
+    protected OpenDistroPortUnificationHandler(OpenDistroSecurityKeyStore odsks, SslHandler providedSSLHandler,
+                                               SSLUtil sslUtils) {
+        this.odsks = odsks;
+        this.providedSSLHandler = providedSSLHandler;
+        this.sslUtils = sslUtils;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        // Will use the first six bytes to detect a protocol.
+        if (in.readableBytes() < 6) {
+            return;
```

Comment:
https://github.com/elastic/elasticsearch/blob/381d7586e4067c1ce0dd840e704b21da9dbeb61b/server/src/main/java/org/elasticsearch/transport/TcpTransport.java#L103 and https://github.com/elastic/elasticsearch/blob/381d7586e4067c1ce0dd840e704b21da9dbeb61b/server/src/main/java/org/elasticsearch/transport/TcpTransport.java#L733-L749

Reply:
In case the first byte is not 0x16, can't we reliably say that it is not TLS?

Reply:
The first byte wont not be sufficient to reliably conclude that the message was TLS or not, also additionally we need to check the first 6 bytes of the message for the custom dual mode client-server handshake

Reply:
In case of TLS should not the first byte be always `0x16`? I understand that you can't reliably say that it is TLS and/or dual mode client-server handshake until all 6 bytes are received, but it should be possible to reliably say that it is *not* using the first byte.

Reply:
The first byte can tell that it is not SSL message in some cases, but there may still be scenarios where we would need to check the first 5 bytes to confirm whether a message is SSL or not. The intention was to always prefer SSL connections when possible. 
Also, we have a custom handshake message exchange which requires us to send at least 6 bytes of data between the transport client and transport server. This custom handshake helps in keeping the connection setup / probe deterministic and quick.
Moreover, the entire logic to check the first 6 bytes only applies during the EIT migration phase (i.e. when dual mode is enabled)
Once the flag is disabled, the DualModeSSLHandler code is never used.